### ISSUE TITLE
Proper path integration for all OS binaries

### ISSUE THREAD

**Issue by Daeraxa** [pulsar-edit/pulsar#187]
### Have you checked for existing feature requests?

- [X] Completed

#### Commands missing from path

See here for [Discord conversation](https://discord.com/channels/992103415163396136/992103415163396139/1047690843496124436).

Basically we should provide a solution to the various packages to make sure that the Pulsar and PPM commands can always be found on the path.

For example if you currently download the Appimage the only way to get the `pulsar` command working is to include it on the path manually even if integrated with appimageinstaller. Same with the tar.gz.

There is already a caveat for macOS in Atom where a command is provided that isn't accessible to other platforms to install the application and apm/ppm commands. (see [conversation](https://discord.com/channels/992103415163396136/992103415163396139/1047697042635505704) and [atom docs](https://pulsar-edit.dev/docs/atom-archive/getting-started/#installing-atom)).

> When you first open Atom, it will try to install the atom and apm commands for use in the terminal. In some cases, Atom might not be able to install these commands because it needs an administrator password. 
... 
> To install the atom and apm commands, run "Window: Install Shell Commands" from the [Command Palette](https://pulsar-edit.dev/getting-started/sections/atom-basics#command-palette), which will prompt you for an administrator password.

There also seems to be a current issue with the Windows binary not installing this to the path which was seemingly handled by Squirrel. [Discord](https://discord.com/channels/992103415163396136/992103415163396139/1047686015638315079).

Now this does work in many places. The .deb file certainly adds `pulsar` to the path and I suspect the `.rpm` and macOS `.dmg` are similar as the `pulsar.sh` file seems to be the one responsible for this.

#### How to fix it?

I think it is important that this gets resolved as currently in the docs it assumes that `pulsar -p` will always work and it seems a bit silly to offer formats where this integration isn't possible without manual steps.
It might be useful simply to expose this command to all platforms and extend the logic to work for all of them. This would allow manual install of the command no matter what system it is run on (and in cases where it doesn't work it will be easier to expand this so that it does). This would also potentially solve the Windows problem and take the guesswork out of the existing macOS case.

There should be a mechanism for this to run automatically on first start as well (maybe check from the config.cson) but it maybe shouldn't run without prompting for those who want to use it entirely as a portable app without leaving any traces. Maybe a first start popup asking to integrate Pulsar on the path with a brief explanation.

### Enter your response below:

Seamless integration of the `pulsar` and `ppm` commands without needing to manually add Pulsar to the path.

### Any alternatives?

Updated installer? A new appimage command for the appimage such as `./Pulsar_1.63.0.appimage --install` that would be the .desktop exec or an option? Use docs to cover this use case? 

None really seem ideal with our supported distributions.

### Other examples:

I think Zed has this option as it doesn't put itself on the path on first install.
Many other applications simply don't come as an appimage and just as OS specific packages which work natively.

**Comment by DeeDeeG** [comment]
The macOS issue is fixed. See https://github.com/pulsar-edit/pulsar/issues/251 and the two linked PRs, which each fixed half of the issues affecting the `pulsar.sh` script on macOS.

**Comment by DeeDeeG** [comment]
Pretty simple fix for Linux: we can symlink to `/usr/bin/ppm` just like we currently symlink to `/usr/bin/pulsar`, at least for .deb and .rpm build targets of electron-builder.

In this file: https://github.com/pulsar-edit/pulsar/blob/5f96f2737d787040ff76999eef6de35d9ee8aeca/script/post-install.sh

Which is run after installing the .deb or .rpm packages, per the electron-builder config: https://github.com/pulsar-edit/pulsar/blob/5f96f2737d787040ff76999eef6de35d9ee8aeca/script/electron-builder.js#L142-L144

---

We could also consider supporting the "Install Shell Commands" command in-app, just like we do on macOS, to workaround any issues with our pacaging, or for folks who have gotten the app as an AppImage or .tar.gz targets, but those are pretty portable/arbitrary where you save them to, and that makes it a lot harder to predict where the real binary will be for us to symlink to...

The AppImage one is kinda unfortunate in that sense, since it's such a smooth user experience otherwise, but I think those users should use `pulsar -p` or roll their own symlink if they don't mind.

Likewise for the .tar.gz folks, I especially think they don't expect automatic system integration from just extracting a compressed archive of the files..

**Comment by DeeDeeG** [comment]
:window: For Windows, I suppose we can try to invoke `setx` with Administrator permissions at some point in the install to add Pulsar's install dir to the PATH? Which would allow running Pulsar from the CLI.

https://stackoverflow.com/a/69383805

Perhaps during the installer: 

And/or we could likewise add this to a Windows version of the "Install Shell Commands" command, which users can manually run in-app.

Portable edition should probably not be put on the PATH, nor used with extensive ppm commands, since it is a self-extracting 7-zip archive, and it takes a super long time to launch... Imagine waiting through that super long extraction process just to run a CLI program...

**Comment by confused-Techie** [comment]
@DeeDeeG I'm personally on board with adding everything to be a variant of "Install Shell Commands" as long as that's clearly documented it feels the best, just to offload the choice and responsibility for users.

And if it's an option for every OS we can add it right into the "System Settings" page like other OS integration features, rather than have it hidden away behind a command. But of course for simplicity the button there can just invoke the command.

But I am really glad to see there is a way to fix, at least quickly for Linux, and is working on MacOS, then looks like it just needs another script for Windows.

An alternative approach for Windows, is from some quick research I was doing a few days ago about `electron-builder` it looks like packaging Windows as an AppX would alleviate the concern about PATH, and allow for some other features Windows Users might expect, like App Uninstallation directly from Settings rather than Control Panel like what currently has to occur

**Comment by DeeDeeG** [comment]
Interesting about AppX...

**Comment by confused-Techie** [comment]
@Daeraxa Now with #604 merged, I can personally say that Pulsar has full proper PATH integration for Windows.

And you mentioned above that macOS should be fine as well, and some versions of the Linux installer.

So really to close this issue out, should we extend the behavior of macOS to Linux to be able to add Pulsar to the PATH manually there as well? I'd love to take a shot at this, but can you think of any examples or possibilities about the code that would have to be implemented itself?

And for posterity:

* [x] Windows
* [x] macOS
* [ ] Linux: AppImage
* [ ] Linux: tar.gz
* [x] Linux: rpm
* [x] Linux: deb

This is looking pretty good, and that we are close to having it fully solved, with obviously a few discrepancies still

**Comment by Daeraxa** [comment]
Really anyone downloading the tar.gz should be (and probably will be) dealing with this themselves, at most we may need to provide instructions on where the binaries are in the file structure.

Appimage is a bit more difficult `pulsar -p` should work out the gate so long as you either integrated the appimage directly (with something like appimagelauncher) or did it yourself - see https://discord.com/channels/992103415163396136/1094684411397681244/1106722578892062811 and https://discord.com/channels/992103415163396136/1094684411397681244/1106728722142331081. Not sure how we could easily expose `ppm` though as you would need to unpack the appimage first. There probably is a way of doing it but very likely beyond me.

**Comment by savetheclocktower** [comment]
I think this might be addressed about as well as we can address it:

* macOS doesn't automatically put them in your `PATH`, but has a command for creating symlinks
* the Windows installer offers you the option of placing `pulsar` and `ppm` in your path
* the `.deb` and `.rpm` Linux versions should put them in your `PATH` automatically
* both the Windows portable download and the Linux `.tar.gz` download can extract to a folder and otherwise act portably; this may be what the user wants, in which case we certainly should not _automatically_ put them in the `PATH` (and at most could introduce a command to create symlinks the way we did on macOS)
* the Linux AppImage is meant to be self-contained; that's the whole point — and we did some AppImage-specific fixes to guarantee that we could both launch the AppImage correctly (and have it route to `pulsar.sh` internally) and use `-p`/`--package` to invoke `ppm`

In addition, the documentation [has a whole page](https://docs.pulsar-edit.dev/getting-started/terminal-commands/) explaining how to put `pulsar` and/or `ppm` into your `PATH`, depending on your OS and how you installed Pulsar.

So I think all that's left to discuss is whether we want to enhance the “Window: Install Shell Commands” command so that it works on Windows portable and Linux `.tar.gz` versions. I'm not opposed to that, but neither do I think it's urgent.

**Comment by confused-Techie** [comment]
@savetheclocktower Yeah I think you're right, this one is addressed and resolved as much as it needs and or should be.

While we could extend out that specific command, really I think we do have valid solutions for every single OS, and am happy closing this out.
